### PR TITLE
engine: Switch std::File to nix and modify interface of storage

### PR DESCRIPTION
**Pull request by uran0sH**
Signed-off-by: uran0sH <huangwenyuu@outlook.com>

### CODE REVIEW

File: server/src/distributed_engine/mod.rs [hunk 1]
```diff
@@ -132,19 +132,18 @@ where
...
-                self.local_storage.create_directory(path)?;
                 return Err(EngineError::from(status));
             }
         }
-        Ok(())
+        self.local_storage.delete_directory(path.clone())
     }
 
-    pub async fn read_dir(&self, path: String) -> Result<Vec<String>, EngineError> {
+    pub async fn read_dir(&self, path: String, size: usize) -> Result<Vec<String>, EngineError> {
```
[the remark this answers is not in the document]
Comment (uran0sH) [reply server/src/distributed_engine/mod.rs:141]: I am confused about the offset of directories. Directories should have a alphabetic sort?

File: server/src/distributed_engine/mod.rs [hunk 2]
```diff
@@ -218,14 +217,23 @@ where
...
-    pub async fn write_file(&self, path: String, data: &[u8]) -> Result<(), EngineError> {
-        // a temporary implementation
-        self.local_storage.write_file(path, data)
+    pub async fn write_file(
+        &self,
+        path: String,
+        data: &[u8],
+        offset: i64,
+    ) -> Result<usize, EngineError> {
+        self.local_storage.write_file(path, data, offset)
```
[the remark this answers is not in the document]
Comment (uran0sH) [reply server/src/distributed_engine/mod.rs:236]: data.len() will be larger than size?

File: server/src/distributed_engine/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (luanshaotong) [reply server/src/distributed_engine/mod.rs:141]: I'm not sure which order to sort directories. We can leave it as preserved, and implement it in the future.

File: server/src/distributed_engine/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (luanshaotong) [reply server/src/distributed_engine/mod.rs:236]: I worry too much.

File: server/src/distributed_engine/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (uran0sH) [reply server/src/distributed_engine/mod.rs:141]: got it

File: server/src/distributed_engine/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (uran0sH) [reply server/src/distributed_engine/mod.rs:236]: https://github.com/labring/sealfs/issues/41 This issue might solve it.

File: server/src/storage_engine/default_engine.rs [hunk 3]
```diff
@@ -213,22 +243,24 @@ impl StorageEngine for DefaultEngine {
...
     }
 
-    fn create_file(&self, path: String) -> Result<(), EngineError> {
-        self.file_attr_db.db.put(&path, "f")?;
+    fn create_file(&self, path: String, mode: Mode) -> Result<i32, EngineError> {
         let local_file_name = generate_local_file_name(&self.root, &path);
+        let oflag = OFlag::O_CREAT | OFlag::O_EXCL;
+        let fd = fcntl::open(local_file_name.as_str(), oflag, mode)?;
+        self.file_attr_db.db.put(&path, "f")?;
+        let path_nlink = path.clone() + "_nlink";
```
Comment (Tricky511) [inline server/src/storage_engine/default_engine.rs:251]: Does `nlink` need to delete after calling `delete_file` ?

File: server/src/distributed_engine/mod.rs [hunk 4]
```diff
@@ -132,19 +132,18 @@ where
...
             if status != 0 {
-                self.local_storage.create_directory(path)?;
                 return Err(EngineError::from(status));
             }
         }
-        Ok(())
+        self.local_storage.delete_directory(path.clone())
     }
 
     pub async fn read_dir(&self, path: String) -> Result<Vec<String>, EngineError> {
```
Comment (luanshaotong) [inline server/src/distributed_engine/mod.rs:141]: So, you decide not to add size and offset to read_dir now?

File: server/src/storage_engine/default_engine.rs (hunk 3 above)
Comment (uran0sH) [reply server/src/storage_engine/default_engine.rs:251]: need it

File: server/src/distributed_engine/mod.rs (hunk 4 above)
Comment (uran0sH) [reply server/src/distributed_engine/mod.rs:141]: yes

**Review by Tricky511 — APPROVED** [review]

**Pull request merged**